ui/apps/platform/src/hooks/useSet.test.ts

```diff
@@ -0,0 +1,52 @@
+import { renderHook, act } from '@testing-library/react-hooks';
+import useSet from './useSet';
+
+test('useSet should accept a starting set', () => {
+    // Membership is based on reference equality
+    const objA = { test: 'test' };
+    const objB = { test: 'test' };
+
+    const { result } = renderHook(() => {
+        const set = useSet(new Set([objA]));
+        return set;
+    });
+
+    expect(result.current.has(objA)).toBeTruthy();
+    expect(result.current.has(objB)).toBeFalsy();
+
+    act(() => {
+        result.current.toggle(objA);
+        result.current.toggle(objB);
+    });
+
+    expect(result.current.has(objA)).toBeFalsy();
+    expect(result.current.has(objB)).toBeTruthy();
+});
+
+test('useSet should correctly toggle items and report their membership', () => {
```

Comment:
As straightforward as it looks, this test actually helped uncover a bug. The `toggle` function in `useSet` was relying on a closure value, which failed to update correctly when multiple toggle calls happened in a row. Using the alternate form of `useState` fixed the issue in the below diff:

```diff
     const [itemSet, setItemSet] = useState(initialSet);

     function toggle(key: T, isOn?: boolean) {
-        const nextSet = new Set(itemSet);
-        const shouldAdd = typeof isOn === 'undefined' ? !itemSet.has(key) : isOn;
-        if (shouldAdd) {
-            nextSet.add(key);
-        } else {
-            nextSet.delete(key);
-        }
-        setItemSet(nextSet);
+        setItemSet((prevSet) => {
+            const nextSet = new Set(prevSet);
+            const shouldAdd = typeof isOn === 'undefined' ? !itemSet.has(key) : isOn;
+            if (shouldAdd) {
+                nextSet.add(key);
+            } else {
+                nextSet.delete(key);
+            }
+            return nextSet;
+        });
     }
```

Reply:
Indeed, please share this at the next team meeting. Return on your investment to write tests!